### PR TITLE
Change performance testing to use benchmark diff tool

### PR DESCRIPTION
**Pull request by matt-kwong**
This takes changes from #10048 and #9945 and adds a tool for Jenkins to output the diff into a comment on the PR that is tested.

### CODE REVIEW

File: tools/jenkins/comment_on_pr.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# This script is invoked by Jenkins to comment $1 on pull requests
+# when triggered by a build
+
+set -e
+
+if [ -z $1 ] || [ -z $JENKINS_OATH_TOKEN ] || [ -z $ghprbPullId ]; then
```
Comment (jtattermusch) [inline tools/jenkins/comment_on_pr.sh]: OATH -> OAUTH

is this actually safe to do? what if somebody runs "set" command and that shows value of all enviroment variables and the value stays in jenkins log forever?

File: tools/jenkins/run_performance.sh [hunk 2]
```diff
@@ -34,4 +34,5 @@ set -ex
 # Enter the gRPC repo root
 cd $(dirname $0)/../..
 
-tools/run_tests/run_performance_tests.py -l c++ node ruby csharp python --netperf --category smoketest
+tools/run_tests/run_microbenchmark.py -c summary --diff_perf origin/$ghprbTargetBranch
```
Comment (jtattermusch) [inline tools/jenkins/run_performance.sh]: qq: does this disable performance tests on PRs tests in all languages and only runs C++?
(not entirely up to date on what run_microbenchmark.py does).

File: tools/jenkins/comment_on_pr.sh (hunk 1 above)
Comment (matt-kwong) [inline tools/jenkins/comment_on_pr.sh]: Done

File: tools/jenkins/run_performance.sh (hunk 2 above)
Comment (matt-kwong) [inline tools/jenkins/run_performance.sh]: I think it does? @ctiller please correct me if I'm wrong. 

The idea behind this change is that the performance PR test results were mostly being ignored. This new performance test will show a diff in performance between a PR and the master branch and this is supposed to offer more valuable data than what the current performance PR test. This was decided by the C core team, so there definitely should be more input before this change is made. Maybe run a separate performance test for wrapper languages?

Comment (jtattermusch) [inline tools/jenkins/run_performance.sh]: ok, so if this really disables non-C/C++ scenarios, I am against it.
It is fine to reduce number of scenarios we are running if there's a problem with the tests running for too long (I think majority of them is C/C++ anyway) but this suite is our only protection against breaking performance tests in master (and as a side effect this works as a de-facto stress test for wrapped languages and has lead to discovering issue in the past) - so removing it seems really off.

I think microbenchmarks are great and so is getting a clear signal about regression on a PR, but they should be added as a new Jenkins job. Killing an existing test suite for OTHER languages in order to save a bit of time setting up a new job is not the right way to do things.

Comment (ctiller) [inline tools/jenkins/run_performance.sh]: Let's move the performance suites into the regular unit test jobs (and run them concurrently if we can).

AFAIK, we have *never* gated a PR on the outcome of this job, which means that the idea of running them in an isolated pool is invalid: we don't use the numbers, we just use the success fail bit.

I'm additionally ok if we disable the running of wrapped language qps tests for a few weeks (even up to a month) at this point in the release cycle until we can get things migrated.

**Comment by jtattermusch** [comment]
On Fri, Mar 10, 2017 at 11:46 PM, Craig Tiller <notifications@github.com>
wrote:

> *@ctiller* commented on this pull request.
> ------------------------------
>
> In tools/jenkins/run_performance.sh
> <https://github.com/grpc/grpc/pull/10079#discussion_r105505057>:
>
> > @@ -34,4 +34,5 @@ set -ex
>  # Enter the gRPC repo root
>  cd $(dirname $0)/../..
>
> -tools/run_tests/run_performance_tests.py -l c++ node ruby csharp python --netperf --category smoketest
> +tools/run_tests/run_microbenchmark.py -c summary --diff_perf origin/$ghprbTargetBranch
>
> Let's move the performance suites into the regular unit test jobs (and run
> them concurrently if we can)
>
AFAIK, we have *never* gated a PR on the outcome of this job, which means
> that the idea of running them in an isolated pool is invalid: we don't use
> the numbers, we just use the success fail bit.
>
I get the point that we actually weren't looking at the numbers (but maybe
we should have and now we could use the same PR commenting technique that's
used in this PR - it would become more interesting once we
actually get to some perf optimization work in wrapped languages - and I'm
hoping it's going to be sooner than later).

We could probably probably turn some of the performance tests into unit
tests, but there's problems with that:
1. The benchmarks should run two separate processes and given how running
tests works in wrapped languages (basically every language has its own
testing framework that runs the tests), it's a non-trivial amount of
work to migrate the tests.
2. Some scenarios (like wrapped_lang_client vs c++ server) are impossible
to run by this technique. Yet in the past some of these scenarios have been
failing and it has revealed some bugs.
3. If we want to reveal potential issues with performance tests on master,
we should run a suite that's as similar as possible on pull requests.
That's the only reliable way to prevent breakages. From my experience, if
some performance scenarios get broken on master, it can take up to weeks
before someone notices (especially if it's just one scenario from many). I
want us to know about breakages before PRs get merged, not after.

To be honest, I don't understand where is the urge to kill existing &
working tests coming from, considering you could just as easily add a new
jenkins job for microbenchmarks (and we could discuss killing existing
tests separately).

I'm additionally ok if we disable the running of wrapped language qps tests
> for a few weeks (even up to a month) at this point in the release cycle
> until we can get things migrated.
>
I would have been fine with that if:
1. owners of wrapped languages would be involved in the decision
2. we had a confirmation that every lang. owner has cycles to actually work
on this (which I doubt).
3. OR we actually first added the unit tests first and then killed the
existing perf PR suite (which seems to be the most reasonable approach btw).

Otherwise I have zero confidence that we would actually add the replacement
perf unit tests in all the languages. We can discuss this offline if you
wish.

**Comment by matt-kwong** [comment]
> is this actually safe to do? what if somebody runs "set" command and that shows value of all enviroment > variables and the value stays in jenkins log forever?
The Jenkins plugin I use to bind the OAuth token masks the secret in the Console Output, so this should prevent accidental leakage. 

@ctiller PTAL - I will change run_performance.sh in a future PR after we decide on a good route for performance testing and I verify the results commenting feature works properly with these changes merged. I had to comment out `subprocess.check_call(['git', 'checkout', args.diff_perf])` to get Jenkins to comment on this PR.

**Comment by matt-kwong** [comment]
@ctiller - The latest results look very noisy and the differences do seem significant. Performance PR took 18 mins https://grpc-testing.appspot.com/job/gRPC_performance_pull_requests/10179/. Could we run this 2-3 times for both master and PR to reduce variance?

**Comment by matt-kwong** [comment]
Sounds good, I'll merge without the change to `run_performance.sh`

**Comment by matt-kwong** [comment]
This PR does not affect any tests and was passing tests earlier, so  merging early.

**Pull request merged**